csrc/fusion_segmenter.cpp

```diff
@@ -1860,34 +1861,36 @@ void eraseInputDistinctRootDomains(Fusion* fusion) {
       }
     }
 
-    NVF_ERROR(new_logical_domain.size() == tv->domain()->contiguity().size());
     TensorDomain* new_td = nullptr;
-
     if (tv->domain()->hasAllocation()) {
       // we need to reorder the logical domain into allocation domain
       // consistently with the mapping from the old TensorView logical domain to
       // its allocation domain
-      const auto& alloc = tv->getAllocationDomain();
-      NVF_ERROR(
-          alloc.size() == logical.size(),
-          "size between logical and alloc doesn't match");
-      const auto rank = alloc.size();
-      std::vector<int64_t> stride_order(rank, -1);
-      for (auto i : c10::irange(rank)) {
-        bool found_match = false;
-        for (auto j : c10::irange(rank)) {
-          if (alloc[i] == logical[j]) {
-            stride_order[j] = static_cast<int64_t>(rank - 1 - i);
-            found_match = true;
-            break;
-          }
-        }
-        NVF_ERROR(
-            found_match,
-            "cannot match IterDomain between allocation domain to logical domain");
+      std::unordered_map<IterDomain*, IterDomain*> old_to_new;
+      for (const auto i : c10::irange(logical.size())) {
+        old_to_new.emplace(logical[i], new_logical_domain[i]);
+      }
+
+      ReplayTransformations replay(tv->getAllocationDomain(), old_to_new);
+      // Without this,
+      // https://github.com/NVIDIA/Fuser/blob/e613929a6c21b3095c8817b01b8f177096a26e60/csrc/transform_iter.cpp#L299
+      // tries to look for root IDs in the map, which shouldn't exist because
+      // the whole purpose of this function is to remove the root domain.
+      replay.setErrorOnFailure(false);
+      // Should we replay.setReplayRFactor(true)? I guess the logical domain
+      // shouldn't be rfactor any more because it becomes the root, but maybe
+      // other IterDomains should inherit rfactor?
+      std::vector<IterDomain*> new_alloc;
+      new_alloc.reserve(tv->getAllocationDomain().size());
+      for (IterDomain* alloc_id : tv->getAllocationDomain()) {
+        new_alloc.push_back(replay.getReplay().at(alloc_id));
       }
       new_td = IrBuilder::create<TensorDomain>(
-          new_logical_domain, stride_order, tv->domain()->contiguity());
+          /*root_domain=*/std::vector<IterDomain*>(),
+          new_logical_domain,
+          new_alloc,
+          /*loop_domain=*/new_alloc,
```

Comment:
Why does the loop domain need to be set as the allocation domain?

Reply:
See https://github.com/NVIDIA/Fuser/pull/3458/commits/cf126e688a84e29cf735ab097bbc2cebbeec5b5a. I did try setting this to new_logical_domain in an earlier version and found that I had to also change TransformReplay::replayCasP to use ReplayTransformations instead of BestEffortReplay. Also, I thought we agreed on trying my best to make loop and allocation the same for fusion inputs, so isResharding can simply examine loop domains?

Reply:
What failure did you see with `replayCasP`?

Conceptually, the loop domain of a fusion input should not matter for anything. The loop domain defines the loop structure of the defining expression of the tensor, and fusion inputs have no defining expression. So, while I think it's okay to set it by the allocation domain, I'm still unclear why it's necessary. That said, resharding makes this analysis a little blurred since fusion inputs are generated by resharding communications.

Reply:
With the following patch on top of this PR
```diff
diff --git a/csrc/fusion_segmenter.cpp b/csrc/fusion_segmenter.cpp
index 14377b5f..6ca194ca 100644
--- a/csrc/fusion_segmenter.cpp
+++ b/csrc/fusion_segmenter.cpp
@@ -1889,7 +1889,7 @@ void eraseInputDistinctRootDomains(Fusion* fusion) {
           /*root_domain=*/std::vector<IterDomain*>(),
           new_logical_domain,
           new_alloc,
-          /*loop_domain=*/new_alloc,
+          /*loop_domain=*/new_logical_domain,
           tv->domain()->contiguity());
     } else {
       new_td = IrBuilder::create<TensorDomain>(
@@ -1912,7 +1912,7 @@ void eraseInputDistinctRootDomains(Fusion* fusion) {
             /*root_domain=*/std::vector<IterDomain*>{},
             /*logical_domain=*/new_logical,
             /*allocation=*/TensorDomain::noReductions(new_td->allocation()),
-            /*loop_domain=*/TensorDomain::noReductions(new_td->allocation()),
+            /*loop_domain=*/new_logical,
             /*contiguity=*/no_red_contiguity);
       } else {
         new_td = IrBuilder::create<TensorDomain>(
diff --git a/csrc/transform_replay.cpp b/csrc/transform_replay.cpp
index 06e15929..a429cb70 100644
--- a/csrc/transform_replay.cpp
+++ b/csrc/transform_replay.cpp
@@ -759,7 +759,7 @@ std::pair<TensorDomain*, int64_t> TransformReplay::replayCasP(
   if (producer->hasAllocation()) {
     auto replay_CasP = BestEffortReplay(
         new_IDs,
-        producer->getLoopDomain(),
+        producer->getMaybeAllocationDomain(),
         logical_map.mapProducerToConsumer(producer->domain(), replayed));
     const auto& p2c_map = replay_CasP.getReplay();
```

InputAllocationIsSplit_Concrete fails with the following
```
Note: Google Test filter = *InputAllocationIsSplit_Concrete
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from AllocationDomainTest
[ RUN      ] AllocationDomainTest.InputAllocationIsSplit_Concrete
[New Thread 0x7fff29a00000 (LWP 2002936)]
[New Thread 0x7fff27a00000 (LWP 2002937)]

Thread 1 "nvfuser_tests" hit Catchpoint 1 (exception thrown), 0x00007ffff795735a in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ffff795735a in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x0000555555b98cca in nvfuser::nvfCheckFail (func=0x5555572c6686 "validateDomainEquivalence", file=0x5555572c5620 "/opt/pytorch/nvfuser/csrc/ir/utils.cpp", line=965, msg=" INTERNAL ASSERT FAILED at \"/opt/pytorch/nvfuser/csrc/ir/utils.cpp\":965, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. dom0 has unreachable IDs. dom0: iS7"...) at /opt/pytorch/nvfuser/csrc/exceptions.cpp:275
#2  0x0000555555b98f54 in nvfuser::nvfErrorFail (func=0x5555572c6686 "validateDomainEquivalence", file=0x5555572c5620 "/opt/pytorch/nvfuser/csrc/ir/utils.cpp", line=965, condMsg=0x5555572c65e0 " INTERNAL ASSERT FAILED at \"/opt/pytorch/nvfuser/csrc/ir/utils.cpp\":965, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. ", userMsg="dom0 has unreachable IDs. dom0: iS7{6}. dom1: ") at /opt/pytorch/nvfuser/csrc/exceptions.cpp:301
#3  0x0000555555f10465 in nvfuser::ir_utils::validateDomainEquivalence (dom0=std::vector of length 1, capacity 1 = {...}, dom1=std::vector of length 0, capacity 2, additional_ids=std::vector of length 0, capacity 0) at /opt/pytorch/nvfuser/csrc/ir/utils.cpp:965
#4  0x0000555555df05f8 in nvfuser::TensorDomain::setAllocationDomain (this=0x7fff3591c200, new_allocation_domain=std::vector of length 0, capacity 2, new_contiguity=std::vector of length 0, capacity 2) at /opt/pytorch/nvfuser/csrc/ir/nodes.cpp:3694
#5  0x00005555563d59db in nvfuser::TransformReplay::replayCasP (consumer=0x7fff2c274cc0, producer=0x7fff2c274940, producer_pos=1, logical_map=..., opt=...) at /opt/pytorch/nvfuser/csrc/transform_replay.cpp:781
#6  0x00005555563d601f in nvfuser::TransformReplay::replayCasP (consumer=0x7fff2c274cc0, producer=0x7fff2c274940, compute_at_axis=-1, opt=...) at /opt/pytorch/nvfuser/csrc/transform_replay.cpp:809
#7  0x00005555563a682f in nvfuser::TensorView::cacheAfter (this=0x7fff2c274940, op_type=nvfuser::LoadStoreOpType::Set, cache_op=nvfuser::CacheOp::Unspecified, propagate_allocation_domain=true, cached_uses=std::vector of length 1, capacity 1 = {...}) at /opt/pytorch/nvfuser/csrc/tensor_view.cpp:1277
#8  0x000055555635d37a in nvfuser::scheduler_utils::cacheInputs (fusion=0x7fff2cdfee00, unroll=true) at /opt/pytorch/nvfuser/csrc/scheduler/utils.cpp:1226
#9  0x00005555562bfd2d in nvfuser::schedulePointwise (fusion=0x7fff2cdfee00, pparams=0x7fff3414a700) at /opt/pytorch/nvfuser/csrc/scheduler/pointwise.cpp:508
#10 0x00005555562c2754 in nvfuser::PointWiseScheduler::schedule (this=0x7fff263c7c70, fusion=0x7fff2cdfee00, params=0x7fff3414a700) at /opt/pytorch/nvfuser/csrc/scheduler/pointwise.cpp:974
#11 0x000055555619eabf in nvfuser::FusionKernelRuntime::compileKernel (this=0x7fff2c9b9980, args=..., sg=0x7fff346a6360) at /opt/pytorch/nvfuser/csrc/runtime/fusion_kernel_runtime.cpp:692
#12 0x000055555619c7fc in nvfuser::FusionKernelRuntime::compileFusionParallel (this=0x7fff2c9b9980, args=...) at /opt/pytorch/nvfuser/csrc/runtime/fusion_kernel_runtime.cpp:365
#13 0x00005555561842a6 in nvfuser::FusionExecutorCache::runFusionWithInputs (this=0x7fffffffcfb0, inputs=..., forced_index_type=std::optional [no contained value], selected_device=std::optional [no contained value]) at /opt/pytorch/nvfuser/csrc/runtime/fusion_executor_cache.cpp:65
#14 0x000055555663e3c5 in nvfuser::AllocationDomainTest_InputAllocationIsSplit_Concrete_Test::TestBody (this=0x7fff2c28d3a0) at /opt/pytorch/nvfuser/tests/cpp/test_allocation_domain.cpp:1435
#15 0x00005555570e8a53 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x7fff2c28d3a0, method=&virtual testing::Test::TestBody(), location=0x555557784a2b "the test body") at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:2612
#16 0x00005555570e1f5f in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x7fff2c28d3a0, method=&virtual testing::Test::TestBody(), location=0x555557784a2b "the test body") at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:2648
#17 0x00005555570bc00a in testing::Test::Run (this=0x7fff2c28d3a0) at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:2687
#18 0x00005555570bcbdb in testing::TestInfo::Run (this=0x7fff378a5280) at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:2836
#19 0x00005555570bd645 in testing::TestSuite::Run (this=0x7fff37820580) at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:3015
#20 0x00005555570ce707 in testing::internal::UnitTestImpl::RunAllTests (this=0x7fff344d0300) at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:5920
#21 0x00005555570e9a68 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x7fff344d0300, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x5555570ce2e8 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x555557785478 "auxiliary test code (environments or event listeners)") at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:2612
#22 0x00005555570e3223 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x7fff344d0300, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x5555570ce2e8 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x555557785478 "auxiliary test code (environments or event listeners)") at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:2648
#23 0x00005555570ccb75 in testing::UnitTest::Run (this=0x5555580beb00 <testing::UnitTest::GetInstance()::instance>) at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest.cc:5484
#24 0x00005555570a748b in RUN_ALL_TESTS () at /opt/pytorch/nvfuser/third_party/googletest/googletest/include/gtest/gtest.h:2317
#25 0x00005555570a7473 in main (argc=1, argv=0x7fffffffd858) at /opt/pytorch/nvfuser/third_party/googletest/googletest/src/gtest_main.cc:64
```

In frame 4, nvfuser::TensorDomain::setAllocationDomain is called with an empty allocation domain. This is because p2c_map at https://github.com/NVIDIA/Fuser/blob/769a4d24f92f9bf0d404bb3643dde15b07770d5e/csrc/transform_replay.cpp#L764 doesn't contain any entries for the allocation IterDomains. I believe that is related to the fact that new_IDs contains only loop IterDomains. 

That being said, I'm also wondering why the pointwise scheduler is trying to propagate allocation at all. Is https://github.com/NVIDIA/Fuser/blob/769a4d24f92f9bf0d404bb3643dde15b07770d5e/csrc/scheduler/utils.cpp#L1229 a bug? I don't understand why we care about propagating allocation domain to cached inputs, which are intermediate TensorViews.

Reply:
I don't think the propagation is necessary. This may be related to a recent change of vectorization validation for padding (#3439).

Reply:
I'm feeling the real issue is `TransformReplay::replayCasP`. I'm not sure why it also modifies the allocation domain of the consumer. Let me look into it.

Reply:
I think there are a couple of things to do. First of all, to support propagation of allocation domains in `replayCasP`, the current `BestEffortReplay` is not enough. It was enough before because the allocation domain was just reordering of logical domains. But I'm not sure if further extending `replayCasP` is a good idea. We added that capability in #372, but maybe they should be kept as separate transformations.

For this particular case, I'm concerned about the use of the allocation domain as the loop domain. It would mean that if an allocation domain was set, its loop domain would be automatically modified. I see that this may be convenient for resharding analyses, but I'm not sure if that would make sense more generally.

I wonder if we should also do `ReplayTransformations` with the loop domain. Previously, we assumed `loop_domain == logical_domain`, so it should have no effect. But if a fusion was already scheduled with some splits/merges and DID, it seems to make sense to replay the loop domain as well here. 

As a first step, maybe we could do something like:

```
std::vector<IterDomain*> new_loop_domain;
if (tv->getAllocationDomain() == tv->getLoopDomain()) {
  new_loop_domain = new_alloc;
} else if (tv->getLogicalDomain() == tv->getLoopDomain()) {
  new_loop_domain = new_logical;
} else {
  // I don't think we should hit here
  NVF_THROW();
}
```

Would this work?

Reply:
Done!